kore/src/Kore/Equation/Application.hs

```diff
@@ -176,6 +210,40 @@ attemptEquation sideCondition termLike equation =
         debugAttemptEquationResult equation result
         return result
 
+    simplifyArgumentWithResult
+        :: MonadTrans t
+        => Predicate (Target variable)
+        -> Maybe (Predicate (Target variable))
+        -> MatchResult (Target variable)
+        -> t simplifier [MatchResult (Target variable)]
+    simplifyArgumentWithResult
+        argument
+        antiLeft
+        (matchPredicate, matchSubstitution)
+      =
+        -- TODO: toMap is unsafe;
+        -- is it impossible to happen (if the frontend generates equations
+        -- correctly), or should we add a new AttemptEquationError case?
+        lift $
+            let toMatchResult Conditional { predicate, substitution } =
+                    (predicate, Substitution.toMap substitution)
+             in Substitution.mergePredicatesAndSubstitutions
+                    sideCondition
+                    ([argument, matchPredicate] <> maybeToList antiLeft)
+                    [mkSubstitution matchSubstitution]
+                    & Logic.observeAllT
+                    & (fmap . fmap) toMatchResult
+
+    mkSubstitution =
+        Substitution.fromMap
+        . Map.fromList
+        . fmap
+            (\(variableName, term)->
+                let variableSort = TermLike.termLikeSort term
+                 in (Variable { variableName, variableSort }, term)
+            )
+        . Map.toList
```

Comment:
This is just a generic conversion from `Map` to `Substitution`, right? Could we move this to an instance of `From` in `Kore.Internal.Substitution`?

We can also use `Map.toAscList` and `Map.fromAscList` here, because this operation can't change the order of the variables.